include/qpid/dispatch/protocol_adaptor.h

```diff
@@ -810,15 +804,32 @@ qdr_link_t *qdr_link_first_attach(qdr_connection_t *conn,
 void qdr_link_second_attach(qdr_link_t *link, qdr_terminus_t *source, qdr_terminus_t *target);
 
 /**
- * qdr_link_detach
+ * qdr_link_detach_received
  *
- * This function is invoked when a link detach arrives.
+ * This function is invoked when a link detach performative arrives from the remote peer. This may the first detach
+ * (peer-initiated link detach) or in response to a detach sent by the router (second detach).
  *
  * @param link The link pointer returned by qdr_link_first_attach or in a FIRST_ATTACH event.
- * @param dt The type of detach that occurred.
  * @param error The link error from the detach frame or 0 if none.
  */
-void qdr_link_detach(qdr_link_t *link, qd_detach_type_t dt, qdr_error_t *error);
+void qdr_link_detach_received(qdr_link_t *link, qdr_error_t *error);
+
+
+/**
+ * qdr_link_closed
+ *
+ * This function is invoked by the adaptor when the link has fully closed. This will be the last call made by the
+ * adaptor for this link. This may be called as a result of a successful detach handshake or due to link loss. This will
+ * also be called during adaptor shutdown on any outstanding links.
+ *
+ * The core may free the qdr_link_t by this call. The adaptor MUST NOT reference the qdr_link_t on return from this
+ * call.
+ *
+ * @param link The link pointer returned by qdr_link_first_attach or in a FIRST_ATTACH event.
+ * @param forced True if the link was closed due to failure or shutdown. False if closed by clean detach handshake.
+ */
+void qdr_link_closed(qdr_link_t *link, bool forced);
```

Comment:
I looked for the usage of this function in the tcp adaptor, say in `close_connection_XSIDE_IO()` - 

```
    if (!!conn->outbound_link) {
        qdr_link_closed(conn->outbound_link, true);
    }
```

In the above usage, we are calling the `qdr_link_closed()` function to close a link. Should the function be called `qdr_link_close()` or `qd_close_link()` instead ? `closed` seems to mean that the link has already been closed

Reply:
> closed seems to mean that the link has already been closed

Indeed that's the idea: the adaptor has detected the link is closed and needs to notify the core (hence **qdr**_link_closed rather than **qd**_link_closed: it's a Core API call not an adaptor call) that the link no longer exists.

I chose **qdr_link_closed** because the function performs essentially the same thing as the exising core function **qdr_connection_closed** except the new one deals with qdr_link_t's not qdr_connection_t's).

You'll see that both adaptors call qdr_connection_close() right before they clean up their connection state objects (qd_tcp_connection_t or qd_connection_t - depending on the adaptor).  The TCP adaptor used to call qdr_link_detach() even though there's not such thing as a detach performative for TCP!  For TCP we basically simulate a network connection drop since there's no AMQP close handshake going on.

See [here](https://github.com/kgiusti/skupper-router/blob/ecf7305f3028c19288f739015dc333957d0cbb49/include/qpid/dispatch/protocol_adaptor.h#L389)  and [here](https://github.com/kgiusti/skupper-router/blob/ecf7305f3028c19288f739015dc333957d0cbb49/include/qpid/dispatch/protocol_adaptor.h#L819)

What do you think?